File: src/2025-02-27-deepclaude37-rust/go/exercises/practice/simple-linked-list/simple_linked_list.go
package linkedlist

import "errors"

// Element represents a node in our linked list
type Element struct {
	Value int
	Next  *Element
}

// List represents a singly linked list with a head pointer and size counter
type List struct {
	head *Element
	size int
}

func New(elements []int) *List {
	l := &List{}
	for _, e := range elements {
		l.Push(e)
	}
	return l
}

func (l *List) Size() int {
	return l.size
}

func (l *List) Push(element int) {
	newElement := &Element{Value: element}
	if l.head == nil {
		l.head = newElement
	} else {
		current := l.head
		for current.Next != nil {
			current = current.Next
		}
		current.Next = newElement
	}
	l.size++
}

func (l *List) Pop() (int, error) {
	if l.head == nil {
		return 0, errors.New("empty list")
	}

	// If there's only one element
	if l.head.Next == nil {
		value := l.head.Value
		l.head = nil
		l.size--
		return value, nil
	}

	// Find the second-to-last element
	current := l.head
	for current.Next.Next != nil {
		current = current.Next
	}

	// Get the value of the last element
	value := current.Next.Value
	// Remove the last element
	current.Next = nil
	l.size--
	return value, nil
}

func (l *List) Array() []int {
	arr := make([]int, 0, l.size)
	current := l.head
	for current != nil {
		arr = append(arr, current.Value)
		current = current.Next
	}
	return arr
}

func (l *List) Reverse() *List {
	reversed := &List{}
	current := l.head
	for current != nil {
		// Prepend to the reversed list (stack-like push)
		reversed.head = &Element{Value: current.Value, Next: reversed.head}
		reversed.size++
		current = current.Next
	}
	return reversed
}

File: src/2025-03-04-12-20-21--deepseek-vr-0304/python/exercises/practice/simple-linked-list/simple_linked_list.py
class EmptyListException(Exception):
    def __init__(self, message):
        super().__init__(message)


class Node:
    def __init__(self, value):
        self._value = value
        self._next = None

    def value(self):
        return self._value

    def next(self):
        return self._next


class LinkedList:
    def __init__(self, values=None):
        self._head = None
        self._length = 0
        if values is not None:
            for value in values:
                self.push(value)

    def __iter__(self):
        current = self._head
        while current:
            yield current.value()
            current = current.next()

    def __len__(self):
        return self._length

    def head(self):
        if self._head is None:
            raise EmptyListException("The list is empty.")
        return self._head

    def push(self, value):
        new_node = Node(value)
        new_node._next = self._head
        self._head = new_node
        self._length += 1

    def pop(self):
        if self._head is None:
            raise EmptyListException("The list is empty.")
        value = self._head.value()
        self._head = self._head.next()
        self._length -= 1
        return value

    def reversed(self):
        values = list(self)
        return LinkedList(reversed(values))
